CUB-related test failures

* Conditions 
CuPy Version          : 7.0.0rc1
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10010
CUDA Driver Version   : 10010
CUDA Runtime Version  : 10010
cuDNN Build Version   : 7402
cuDNN Version         : 7402
NCCL Build Version    : None
NCCL Runtime Version  : None

* Code to reproduce
run pytest with / without CUB enabled

* Error messages, stack traces, or logs

The following test cases give me an error when running pytest with CUB_DISABALED=0, but pass with CUB_DISABLED=1. I haven't looked into specific cases, but based on the test names I would guess it is a combination of a few possible things:

1.) reduction in the presence of NaN values
2.) reduction of zero-sized arrays
3.) errors in some axis-specific reduction cases

```


 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:39 TestArrayReduction.test_max_nan
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:51 TestArrayReduction.test_max_nan_imag
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:45 TestArrayReduction.test_max_nan_real
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:87 TestArrayReduction.test_min_nan
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:99 TestArrayReduction.test_min_nan_imag
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:93 TestArrayReduction.test_min_nan_real
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_20.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_20.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_49.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_49.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_136.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_136.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:289 TestEinSumUnaryOperation_param_136.test_einsum_unary_views
 - tests/cupy_tests/linalg_tests/test_norms.py:18 TestTrace.test_external_trace
 - tests/cupy_tests/linalg_tests/test_norms.py:12 TestTrace.test_trace
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_9.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_10.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_11.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_12.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_13.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_14.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_9.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_10.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_11.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_12.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_13.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_14.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_15.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_16.test_pad
 - tests/cupy_tests/sorting_tests/test_search.py:23 TestSearch.test_argmax_nan
 - tests/cupy_tests/sorting_tests/test_search.py:65 TestSearch.test_argmax_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:89 TestSearch.test_argmin_nan
 - tests/cupy_tests/sorting_tests/test_search.py:137 TestSearch.test_argmin_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:327 TestNanArgMin.test_nanargmin_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:417 TestNanArgMax.test_nanargmax_zero_size
```

cc @leofang and @anaruse as fellow CUB contributors here

ref: #2579 which should help avoid similar issues in the future

Hmm, I'm trying to reproduce but currently unable to build CuPy master with CUB 1.8.0, maybe I'm missing something...

`cupy/core/include/cupy/complex/arithmetic.h:262:80: error: ‘double thrust::abs(const thrust::complex<double>&)’ is not declared in ‘thrust’`

<details>

```
building 'cupy.cuda.cub' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -D_FORCE_INLINES=1 -I/usr/local/cuda/include -I/home/kenichi/Development/cupy/install/../cupy/core/include -I/home/kenichi/Development/cub -I/home/kenichi/.pyenv/versions/3.7.0/envs/local-3.7.0/include -I/home/kenichi/.pyenv/versions/3.7.0/include/python3.7m -c cupy/cuda/cub.cpp -o build/temp.linux-x86_64-3.7/cupy/cuda/cub.o -fopenmp
    NVCC options: ['--generate-code=arch=compute_52,code=sm_52', '-O2', '--compiler-options="-fPIC"', '--std=c++11']
/usr/local/cuda/bin/nvcc -D_FORCE_INLINES=1 -I/usr/local/cuda/include -I/home/kenichi/Development/cupy/install/../cupy/core/include -I/home/kenichi/Development/cub -I/home/kenichi/.pyenv/versions/3.7.0/envs/local-3.7.0/include -I/home/kenichi/.pyenv/versions/3.7.0/include/python3.7m -c cupy/cuda/cupy_cub.cu -o build/temp.linux-x86_64-3.7/cupy/cuda/cupy_cub.o --generate-code=arch=compute_52,code=sm_52 -O2 --compiler-options="-fPIC" --std=c++11                                                                                                                            /home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/arithmetic.h:258:78: error: ‘float thrust::abs(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/arithmetic.h:262:80: error: ‘double thrust::abs(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/cexp.h:169:91: error: ‘thrust::complex<double> thrust::exp(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/cexpf.h:149:89: error: ‘thrust::complex<float> thrust::exp(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/clog.h:192:91: error: ‘thrust::complex<double> thrust::log(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/clogf.h:188:89: error: ‘thrust::complex<float> thrust::log(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccosh.h:195:2: error: ‘thrust::complex<double> thrust::cos(const thrust::complex<double>&)’ is not declared in ‘thrust’
     const thrust::complex<double>& z) {
  ^
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccosh.h:201:2: error: ‘thrust::complex<double> thrust::cosh(const thrust::complex<double>&)’ is not declared in ‘thrust’
     const thrust::complex<double>& z) {
  ^
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccoshf.h:126:89: error: ‘thrust::complex<float> thrust::cos(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccoshf.h:131:90: error: ‘thrust::complex<float> thrust::cosh(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinh.h:183:91: error: ‘thrust::complex<double> thrust::sin(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinh.h:188:92: error: ‘thrust::complex<double> thrust::sinh(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinhf.h:124:89: error: ‘thrust::complex<float> thrust::sin(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinhf.h:129:90: error: ‘thrust::complex<float> thrust::sinh(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ctanh.h:182:91: error: ‘thrust::complex<double> thrust::tan(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ctanhf.h:107:89: error: ‘thrust::complex<float> thrust::tan(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csqrt.h:140:92: error: ‘thrust::complex<double> thrust::sqrt(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csqrtf.h:137:90: error: ‘thrust::complex<float> thrust::sqrt(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:697:92: error: ‘thrust::complex<double> thrust::acos(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:702:92: error: ‘thrust::complex<double> thrust::asin(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:708:92: error: ‘thrust::complex<double> thrust::atan(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:725:93: error: ‘thrust::complex<double> thrust::atanh(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:411:90: error: ‘thrust::complex<float> thrust::acos(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:416:90: error: ‘thrust::complex<float> thrust::asin(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:422:90: error: ‘thrust::complex<float> thrust::atan(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:439:91: error: ‘thrust::complex<float> thrust::atanh(const thrust::complex<float>&)’ is not declared in ‘thrust’
error: command '/usr/local/cuda/bin/nvcc' failed with exit status 1
```

--- a/cupy/core/_routines_math.pyx
+++ b/cupy/core/_routines_math.pyx
@@ -73,7 +73,7 @@ cdef ndarray _ndarray_prod(ndarray self, axis, dtype, out, keepdims):
 
 
 cdef ndarray _ndarray_sum(ndarray self, axis, dtype, out, keepdims):
-    if cupy.cuda.cub_enabled:
+    if cupy.cuda.cub_enabled and self._c_contiguous:
         if cub.can_use_device_reduce(cub.CUPY_CUB_SUM, self.dtype, self.ndim,
                                      axis, dtype):
             return cub.device_reduce(self, cub.CUPY_CUB_SUM, out=out,

--- a/cupy/core/_routines_statistics.pyx
+++ b/cupy/core/_routines_statistics.pyx
@@ -13,7 +13,7 @@ if cupy.cuda.cub_enabled:
 
 
 cdef ndarray _ndarray_max(ndarray self, axis, out, dtype, keepdims):
-    if cupy.cuda.cub_enabled:
+    if cupy.cuda.cub_enabled and self._c_contiguous:
         if cub.can_use_device_reduce(cub.CUPY_CUB_MAX, self.dtype, self.ndim,
                                      axis, dtype):
             return cub.device_reduce(self, cub.CUPY_CUB_MAX, out=out,
@@ -26,7 +26,7 @@ cdef ndarray _ndarray_max(ndarray self, axis, out, dtype, keepdims):
 
 
 cdef ndarray _ndarray_min(ndarray self, axis, out, dtype, keepdims):
-    if cupy.cuda.cub_enabled:
+    if cupy.cuda.cub_enabled and self._c_contiguous:
         if cub.can_use_device_reduce(cub.CUPY_CUB_MIN, self.dtype, self.ndim,
                                      axis, dtype):
             return cub.device_reduce(self, cub.CUPY_CUB_MIN, out=out,
@@ -40,7 +40,7 @@ cdef ndarray _ndarray_min(ndarray self, axis, out, dtype, keepdims):
 
 # TODO(leofang): this signature is incompatible with NumPy!
 cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
-    if cupy.cuda.cub_enabled:
+    if cupy.cuda.cub_enabled and self._c_contiguous:
         # Note that the NumPy signature of argmax only has axis and out, so we
         # need to disable the rest. Moreover, to be compatible with NumPy, axis
         # can only be None or integers
@@ -54,7 +54,7 @@ cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
 
 # TODO(leofang): this signature is incompatible with NumPy!
 cdef ndarray _ndarray_argmin(ndarray self, axis, out, dtype, keepdims):
-    if cupy.cuda.cub_enabled:
+    if cupy.cuda.cub_enabled and self._c_contiguous:
         # Note that the NumPy signature of argmax only has axis and out, so we
         # need to disable the rest. Moreover, to be compatible with NumPy, axis
         # can only be None or integers

--- a/cupy/core/_routines_statistics.pyx
+++ b/cupy/core/_routines_statistics.pyx
@@ -44,7 +44,7 @@ cdef ndarray _ndarray_argmax(ndarray self, axis, out, dtype, keepdims):
         # Note that the NumPy signature of argmax only has axis and out, so we
         # need to disable the rest. Moreover, to be compatible with NumPy, axis
         # can only be None or integers
-        if axis is None and cub.can_use_device_reduce(
+        if cub.can_use_device_reduce(
                 cub.CUPY_CUB_ARGMAX, self.dtype, self.ndim, axis, None):
             return cub.device_reduce(self, cub.CUPY_CUB_ARGMAX, out=out,
                                      keepdims=False)
@@ -58,7 +58,7 @@ cdef ndarray _ndarray_argmin(ndarray self, axis, out, dtype, keepdims):
         # Note that the NumPy signature of argmax only has axis and out, so we
         # need to disable the rest. Moreover, to be compatible with NumPy, axis
         # can only be None or integers
-        if axis is None and cub.can_use_device_reduce(
+        if cub.can_use_device_reduce(
                 cub.CUPY_CUB_ARGMIN, self.dtype, self.ndim, axis, None):
             return cub.device_reduce(self, cub.CUPY_CUB_ARGMIN, out=out,
                                      keepdims=False)

--- a/cupy/core/include/cupy/complex.cuh
+++ b/cupy/core/include/cupy/complex.cuh
@@ -24,9 +24,11 @@ using thrust::asin;
 using thrust::acos;
 using thrust::atan;
 
-template<typename T> __device__ bool isnan(complex<T> x) {
+template<typename T>
+__host__ __device__ bool isnan(complex<T> x) {
     return isnan(x.real()) || isnan(x.imag());
 }
+
 template<typename T> __device__ bool isinf(complex<T> x) {
     return isinf(x.real()) || isinf(x.imag());
 }
@@ -49,7 +51,8 @@ template<typename T> __device__ complex<T> expm1(complex<T> x) {
     return y;
 }
 
-template<typename T> __device__ complex<T> min(complex<T> x, complex<T> y) {
+template<typename T>
+__host__ __device__ complex<T> min(complex<T> x, complex<T> y) {
     if (isnan(x)) {
         return y;
     } else if (isnan(y)) {
@@ -64,7 +67,9 @@ template<typename T> __device__ complex<T> min(complex<T> x, complex<T> y) {
         return y;
     }
 }
-template<typename T> __device__ complex<T> max(complex<T> x, complex<T> y) {
+
+template<typename T>
+__host__ __device__ complex<T> max(complex<T> x, complex<T> y) {
     if (isnan(x)) {
         return y;
     } else if (isnan(y)) {

--- a/cupy/core/include/cupy/complex/arithmetic.h
+++ b/cupy/core/include/cupy/complex/arithmetic.h
@@ -118,7 +118,7 @@ __device__ inline complex<ValueType> operator/(const ValueType& lhs,
  * a > c or a == c and b > d. --- */
 
 template <typename ValueType>
-__device__ inline bool operator<(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator<(const complex<ValueType>& lhs,
                                  const complex<ValueType>& rhs) {
   if (lhs == rhs)
   {
@@ -136,7 +136,7 @@ __device__ inline bool operator<(const complex<ValueType>& lhs,
 }
 
 template <typename ValueType>
-__device__ inline bool operator<=(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator<=(const complex<ValueType>& lhs,
                                   const complex<ValueType>& rhs) {
   if (lhs == rhs)
   {
@@ -151,7 +151,7 @@ __device__ inline bool operator<=(const complex<ValueType>& lhs,
 }
 
 template <typename ValueType>
-__device__ inline bool operator>(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator>(const complex<ValueType>& lhs,
                                  const complex<ValueType>& rhs) {
   if (lhs == rhs)
   {
@@ -163,7 +163,7 @@ __device__ inline bool operator>(const complex<ValueType>& lhs,
 }
 
 template <typename ValueType>
-__device__ inline bool operator>=(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator>=(const complex<ValueType>& lhs,
                                   const complex<ValueType>& rhs) {
   if (lhs == rhs || lhs > rhs)
   {
@@ -175,49 +175,49 @@ __device__ inline bool operator>=(const complex<ValueType>& lhs,
 }
 
 template <typename ValueType>
-__device__ inline bool operator<(const ValueType& lhs,
+__host__ __device__ inline bool operator<(const ValueType& lhs,
                                  const complex<ValueType>& rhs) {
     return complex<ValueType>(lhs) < rhs;
 }
 
 template <typename ValueType>
-__device__ inline bool operator>(const ValueType& lhs,
+__host__ __device__ inline bool operator>(const ValueType& lhs,
                                  const complex<ValueType>& rhs) {
     return complex<ValueType>(lhs) > rhs;
 }
 
 template <typename ValueType>
-__device__ inline bool operator<(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator<(const complex<ValueType>& lhs,
                                  const ValueType& rhs) {
     return lhs < complex<ValueType>(rhs);
 }
 
 template <typename ValueType>
-__device__ inline bool operator>(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator>(const complex<ValueType>& lhs,
                                  const ValueType& rhs) {
     return lhs > complex<ValueType>(rhs);
 }
 
 template <typename ValueType>
-__device__ inline bool operator<=(const ValueType& lhs,
+__host__ __device__ inline bool operator<=(const ValueType& lhs,
                                   const complex<ValueType>& rhs) {
     return complex<ValueType>(lhs) <= rhs;
 }
 
 template <typename ValueType>
-__device__ inline bool operator>=(const ValueType& lhs,
+__host__ __device__ inline bool operator>=(const ValueType& lhs,
                                   const complex<ValueType>& rhs) {
     return complex<ValueType>(lhs) >= rhs;
 }
 
 template <typename ValueType>
-__device__ inline bool operator<=(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator<=(const complex<ValueType>& lhs,
                                   const ValueType& rhs) {
     return lhs <= complex<ValueType>(rhs);
 }
 
 template <typename ValueType>
-__device__ inline bool operator>=(const complex<ValueType>& lhs,
+__host__ __device__ inline bool operator>=(const complex<ValueType>& lhs,
                                   const ValueType& rhs) {
     return lhs >= complex<ValueType>(rhs);
 }

--- a/cupy/cuda/cub.pxd
+++ b/cupy/cuda/cub.pxd
@@ -1,4 +1,4 @@
-cpdef enum:
+cpdef enum cupy_cub_op:
     CUPY_CUB_SUM = 0
     CUPY_CUB_MIN = 1
     CUPY_CUB_MAX = 2

--- a/cupy/cuda/cub.pyx
+++ b/cupy/cuda/cub.pyx
@@ -88,7 +88,7 @@ cpdef _preprocess_array(ndarray arr, axis, bint keepdims):
     return out_shape, contiguous_size
 
 
-def device_reduce(ndarray x, int op, out=None, bint keepdims=False):
+def device_reduce(ndarray x, op, out=None, bint keepdims=False):
     cdef ndarray y, z
     cdef ndarray ws
     cdef int dtype_id, ndim_out, kv_bytes
@@ -106,6 +106,9 @@ def device_reduce(ndarray x, int op, out=None, bint keepdims=False):
     if op < CUPY_CUB_SUM or op > CUPY_CUB_ARGMAX:
         raise ValueError("only CUPY_CUB_SUM, CUPY_CUB_MIN, CUPY_CUB_MAX, "
                          "CUPY_CUB_ARGMIN, and CUPY_CUB_ARGMAX are supported.")
+    if x.size == 0 and op != CUPY_CUB_SUM:
+        raise ValueError('zero-size array to reduction operation {} which has '
+                         'no identity'.format(op.name))
     x = _internal_ascontiguousarray(x)
     if CUPY_CUB_SUM <= op <= CUPY_CUB_MAX:
         y = ndarray((), x.dtype)
@@ -137,7 +140,7 @@ def device_reduce(ndarray x, int op, out=None, bint keepdims=False):
     return y
 
 
-def device_segmented_reduce(ndarray x, int op, axis, out=None,
+def device_segmented_reduce(ndarray x, op, axis, out=None,
                             bint keepdims=False):
     # if import at the top level, a segfault would happen when import cupy!
     from cupy.creation.ranges import arange
@@ -156,6 +159,9 @@ def device_segmented_reduce(ndarray x, int op, axis, out=None,
     if op < CUPY_CUB_SUM or op > CUPY_CUB_MAX:
         raise ValueError("only CUPY_CUB_SUM, CUPY_CUB_MIN, and CUPY_CUB_MAX "
                          "are supported.")
+    if x.size == 0 and op != CUPY_CUB_SUM:
+        raise ValueError('zero-size array to reduction operation {} which has '
+                         'no identity'.format(op.name))
 
     # prepare input
     out_shape, contiguous_size = _preprocess_array(x, axis, keepdims)
@@ -165,6 +171,11 @@ def device_segmented_reduce(ndarray x, int op, axis, out=None,
     if out is not None and out.shape != out_shape:
         raise ValueError(
             "output parameter for reduction operation has the wrong shape")
+    if x.size == 0:  # for CUPY_CUB_SUM
+        if out is not None:
+            y = out
+        y[...] = 0
+        return y
     n_segments = x.size//contiguous_size
     # CUB internally use int for offset...
     offset = arange(0, x.size+1, contiguous_size, dtype=numpy.int32)

--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -48,12 +48,6 @@ template <> struct NumericTraits<complex<double>> : BaseTraits<FLOATING_POINT, t
    TODO(leofang): support half precision?
 */
 
-// 1. The function name is appended "cupy_" to avoid confusion and name collision
-// 2. std::isnan() cannot be used here as <cmath> cannot be included
-// 3. This is used for floating-point types (float, double)
-template <typename T>
-__host__ __device__ __forceinline__ bool cupy_isnan(const T &a) { return a != a; }
-
 //
 // Max()
 //
@@ -63,8 +57,8 @@ template <>
 __host__ __device__ __forceinline__ float Max::operator()(const float &a, const float &b) const
 {
     // NumPy behavior: NaN is always chosen!
-    if (cupy_isnan(a)) {return a;}
-    else if (cupy_isnan(b)) {return b;}
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
     else {return CUB_MAX(a, b);}
 }
 
@@ -73,8 +67,8 @@ template <>
 __host__ __device__ __forceinline__ double Max::operator()(const double &a, const double &b) const
 {
     // NumPy behavior: NaN is always chosen!
-    if (cupy_isnan(a)) {return a;}
-    else if (cupy_isnan(b)) {return b;}
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
     else {return CUB_MAX(a, b);}
 }
 
@@ -111,8 +105,8 @@ template <>
 __host__ __device__ __forceinline__ float Min::operator()(const float &a, const float &b) const
 {
     // NumPy behavior: NaN is always chosen!
-    if (cupy_isnan(a)) {return a;}
-    else if (cupy_isnan(b)) {return b;}
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
     else {return CUB_MIN(a, b);}
 }
 
@@ -121,8 +115,8 @@ template <>
 __host__ __device__ __forceinline__ double Min::operator()(const double &a, const double &b) const
 {
     // NumPy behavior: NaN is always chosen!
-    if (cupy_isnan(a)) {return a;}
-    else if (cupy_isnan(b)) {return b;}
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
     else {return CUB_MIN(a, b);}
 }
 
@@ -160,9 +154,9 @@ __host__ __device__ __forceinline__ KeyValuePair<int, float> ArgMax::operator()(
     const KeyValuePair<int, float> &a,
     const KeyValuePair<int, float> &b) const
 {
-    if (cupy_isnan(a.value))
+    if (isnan(a.value))
         return a;
-    else if (cupy_isnan(b.value))
+    else if (isnan(b.value))
         return b;
     else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
         return b;
@@ -176,9 +170,9 @@ __host__ __device__ __forceinline__ KeyValuePair<int, double> ArgMax::operator()
     const KeyValuePair<int, double> &a,
     const KeyValuePair<int, double> &b) const
 {
-    if (cupy_isnan(a.value))
+    if (isnan(a.value))
         return a;
-    else if (cupy_isnan(b.value))
+    else if (isnan(b.value))
         return b;
     else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
         return b;
@@ -228,9 +222,9 @@ __host__ __device__ __forceinline__ KeyValuePair<int, float> ArgMin::operator()(
     const KeyValuePair<int, float> &a,
     const KeyValuePair<int, float> &b) const
 {
-    if (cupy_isnan(a.value))
+    if (isnan(a.value))
         return a;
-    else if (cupy_isnan(b.value))
+    else if (isnan(b.value))
         return b;
     else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
         return b;
@@ -244,9 +238,9 @@ __host__ __device__ __forceinline__ KeyValuePair<int, double> ArgMin::operator()
     const KeyValuePair<int, double> &a,
     const KeyValuePair<int, double> &b) const
 {
-    if (cupy_isnan(a.value))
+    if (isnan(a.value))
         return a;
-    else if (cupy_isnan(b.value))
+    else if (isnan(b.value))
         return b;
     else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
         return b;

--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -6,13 +6,13 @@
 
 using namespace cub;
 
-// Minimum boilerplate to support complex numbers in sum(), min(), and max():
+/* ------------------------------------ Minimum boilerplate to support complex numbers ------------------------------------ */
 // - This works only because all data fields in the *Traits struct are not
 //   used in <cub/device/device_reduce.cuh>.
-// - DO NOT USE THIS STUB for supporting CUB sorting!!!!!!
 // - The Max() and Lowest() below are chosen to comply with NumPy's lexical
 //   ordering; note that std::numeric_limits<T> does not support complex
 //   numbers as in general the comparison is ill defined.
+// - DO NOT USE THIS STUB for supporting CUB sorting!!!!!!
 template <>
 struct FpLimits<complex<float>>
 {
@@ -39,7 +39,253 @@ struct FpLimits<complex<double>>
 
 template <> struct NumericTraits<complex<float>>  : BaseTraits<FLOATING_POINT, true, false, unsigned int, complex<float>> {};
 template <> struct NumericTraits<complex<double>> : BaseTraits<FLOATING_POINT, true, false, unsigned long long, complex<double>> {};
-// end of boilerplate
+/* ------------------------------------ end of boilerplate ------------------------------------ */
+
+
+/* ------------------------------------ "Patches" to CUB ------------------------------------
+   These stubs are needed because CUB does not handle NaNs properly, while NumPy has certain
+   behaviors with which we must comply.
+   TODO(leofang): support half precision?
+*/
+
+// 1. The function name is appended "cupy_" to avoid confusion and name collision
+// 2. std::isnan() cannot be used here as <cmath> cannot be included
+// 3. This is used for floating-point types (float, double)
+template <typename T>
+__host__ __device__ __forceinline__ bool cupy_isnan(const T &a) { return a != a; }
+
+//
+// Max()
+//
+
+// specialization for float for handling NaNs
+template <>
+__host__ __device__ __forceinline__ float Max::operator()(const float &a, const float &b) const
+{
+    // NumPy behavior: NaN is always chosen!
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MAX(a, b);}
+}
+
+// specialization for double for handling NaNs
+template <>
+__host__ __device__ __forceinline__ double Max::operator()(const double &a, const double &b) const
+{
+    // NumPy behavior: NaN is always chosen!
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MAX(a, b);}
+}
+
+// specialization for complex<float> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ complex<float> Max::operator()(const complex<float> &a, const complex<float> &b) const
+{
+    // - TODO(leofang): just call max() here when the bug in cupy/complex.cuh is fixed
+    // - NumPy behavior: If both a and b contain NaN, the first argument is chosen
+    // - isnan() and max() are defined in cupy/complex.cuh
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
+    else {return max(a, b);}
+}
+
+// specialization for complex<double> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ complex<double> Max::operator()(const complex<double> &a, const complex<double> &b) const
+{
+    // - TODO(leofang): just call max() here when the bug in cupy/complex.cuh is fixed
+    // - NumPy behavior: If both a and b contain NaN, the first argument is chosen
+    // - isnan() and max() are defined in cupy/complex.cuh
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
+    else {return max(a, b);}
+}
+
+//
+// Min()
+//
+
+// specialization for float for handling NaNs
+template <>
+__host__ __device__ __forceinline__ float Min::operator()(const float &a, const float &b) const
+{
+    // NumPy behavior: NaN is always chosen!
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MIN(a, b);}
+}
+
+// specialization for double for handling NaNs
+template <>
+__host__ __device__ __forceinline__ double Min::operator()(const double &a, const double &b) const
+{
+    // NumPy behavior: NaN is always chosen!
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MIN(a, b);}
+}
+
+// specialization for complex<float> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ complex<float> Min::operator()(const complex<float> &a, const complex<float> &b) const
+{
+    // - TODO(leofang): just call min() here when the bug in cupy/complex.cuh is fixed
+    // - NumPy behavior: If both a and b contain NaN, the first argument is chosen
+    // - isnan() and min() are defined in cupy/complex.cuh
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
+    else {return min(a, b);}
+}
+
+// specialization for complex<double> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ complex<double> Min::operator()(const complex<double> &a, const complex<double> &b) const
+{
+    // - TODO(leofang): just call min() here when the bug in cupy/complex.cuh is fixed
+    // - NumPy behavior: If both a and b contain NaN, the first argument is chosen
+    // - isnan() and min() are defined in cupy/complex.cuh
+    if (isnan(a)) {return a;}
+    else if (isnan(b)) {return b;}
+    else {return min(a, b);}
+}
+
+//
+// ArgMax()
+//
+
+// specialization for float for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, float> ArgMax::operator()(
+    const KeyValuePair<int, float> &a,
+    const KeyValuePair<int, float> &b) const
+{
+    if (cupy_isnan(a.value))
+        return a;
+    else if (cupy_isnan(b.value))
+        return b;
+    else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for double for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, double> ArgMax::operator()(
+    const KeyValuePair<int, double> &a,
+    const KeyValuePair<int, double> &b) const
+{
+    if (cupy_isnan(a.value))
+        return a;
+    else if (cupy_isnan(b.value))
+        return b;
+    else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for complex<float> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, complex<float>> ArgMax::operator()(
+    const KeyValuePair<int, complex<float>> &a,
+    const KeyValuePair<int, complex<float>> &b) const
+{
+    if (isnan(a.value))
+        return a;
+    else if (isnan(b.value))
+        return b;
+    else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for complex<double> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, complex<double>> ArgMax::operator()(
+    const KeyValuePair<int, complex<double>> &a,
+    const KeyValuePair<int, complex<double>> &b) const
+{
+    if (isnan(a.value))
+        return a;
+    else if (isnan(b.value))
+        return b;
+    else if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+//
+// ArgMin()
+//
+
+// specialization for float for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, float> ArgMin::operator()(
+    const KeyValuePair<int, float> &a,
+    const KeyValuePair<int, float> &b) const
+{
+    if (cupy_isnan(a.value))
+        return a;
+    else if (cupy_isnan(b.value))
+        return b;
+    else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for double for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, double> ArgMin::operator()(
+    const KeyValuePair<int, double> &a,
+    const KeyValuePair<int, double> &b) const
+{
+    if (cupy_isnan(a.value))
+        return a;
+    else if (cupy_isnan(b.value))
+        return b;
+    else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for complex<float> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, complex<float>> ArgMin::operator()(
+    const KeyValuePair<int, complex<float>> &a,
+    const KeyValuePair<int, complex<float>> &b) const
+{
+    if (isnan(a.value))
+        return a;
+    else if (isnan(b.value))
+        return b;
+    else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+
+// specialization for complex<double> for handling NaNs
+template <>
+__host__ __device__ __forceinline__ KeyValuePair<int, complex<double>> ArgMin::operator()(
+    const KeyValuePair<int, complex<double>> &a,
+    const KeyValuePair<int, complex<double>> &b) const
+{
+    if (isnan(a.value))
+        return a;
+    else if (isnan(b.value))
+        return b;
+    else if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key)))
+        return b;
+    else
+        return a;
+}
+/* ------------------------------------ End of "patches" ------------------------------------ */
 
 
 //

